### PR TITLE
Fall back to path when comparing tracks

### PR DESCRIPTION
**Pull request by lbonn**
Tracks will be sorted by ascending path if neither album, disc and track id can help.

This is helpful for browsing a loose collection of tracks tagged under the same album.

Similar to #219

I've tested the changes in an emulator, with navidrome as a server.

### CODE REVIEW

**Comment by ogarcia** [comment]
@lbonn thanks for your PR but you must sign your commit, see [CONTRIBUTING](CONTRIBUTING.md).

**Comment by lbonn** [comment]
@ogarcia I tried to follow the insctructions and used `--signoff` to acknowledge the contribution will be licensed as GPLv3.

Do you mean cryptographically signed?
Or the "Signed-off-by:" must be followed by a real name?

**Comment by Maxmystere** [comment]
image
You should have this when commit is signed
(Github auto sign commits when done with the web text editor if you don't wanna bother yourself with recommiting from repo)

**Comment by ogarcia** [comment]
@lbonn [This](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/about-commit-signature-verification)

I just update CONTRIBUTING to clarify this point :wink:

**Comment by lbonn** [comment]
Thanks for the clarification, I will try to do that by the end of the week-end.

But then I realized I might have taken the wrong approach, and that the sorting feature seems to be intended as a last resort for broken ordering from the server and that it should be first fixed there if possible.
For example, it also sorts "Most Played" and "Recently Played" albums by their names (though, we could probably fix that as well)

But we can still merge this if you think it's a good addition.

**Review by nitehu — APPROVED** [review]
I think the change is OK, if it fixes the ordering for some users... With my test servers the track comparision usually has a result, so it won't go comparing the paths unless I change the result in debug.

**Comment by lbonn** [comment]
Ok, re-pushed after a gpg signature.

File: ultrasonic/src/main/java/org/moire/ultrasonic/util/EntryByDiscAndTrackComparator.java [hunk 1]
```diff
@@ -33,7 +35,14 @@ public int compare(MusicDirectory.Entry x, MusicDirectory.Entry y)
...
 
-		return compare(trackX == null ? 0 : trackX, trackY == null ? 0 : trackY);
+		int trackComparison = compare(trackX == null ? 0 : trackX, trackY == null ? 0 : trackY);
+
+		if (trackComparison != 0)
+		{
+			return trackComparison;
+		}
+
+		return compare(pathX == null ? "" : pathX, pathY == null ? "" : pathY);
```
Comment (ogarcia) [inline ultrasonic/src/main/java/org/moire/ultrasonic/util/EntryByDiscAndTrackComparator.java:45]: I have a doubt with this. `compare` is a function that compares two _longs_, but `pathX` and `pathY` are strings. I think that this is a `TypeError`

Comment (lbonn) [reply ultrasonic/src/main/java/org/moire/ultrasonic/util/EntryByDiscAndTrackComparator.java:45]: It's overloaded, there is a `private static int compare(String a, String b)` on line 53

**Pull request merged**